### PR TITLE
Use FFI::DataConverter to define native handle and add type/nil checks

### PR DESCRIPTION
**Pull request by Adithya-copart**
Alternate take on #126

### CODE REVIEW

**Comment by Adithya-copart** [comment]
Added 26beb479dcd87a570a248cd307cdd8633694f6b5 as there are two calls to `to_native` as mentioned in https://github.com/ffi/ffi/issues/807

**Comment by thijsc** [comment]
Looking good, what's still draft about it?

**Comment by Adithya-copart** [comment]
Hey @thijsc ,

I'm thinking to add more classes that use `FFI::DataConverter` so that references like `:pointer, :int32 etc` can be replaced with these classes. It would be a refactor like the `Handle` class. What are your thoughts?

Should we rather merge the `Handle` class first?

Also, this PR has a `Rdkafka::BaseError` that could be used as the base class of all errors that can be raised in the library. I inherited the new errors introduced in this PR from this class. Is this an approach you're fine with?

File: spec/rdkafka/producer_spec.rb [hunk 1]
```diff
@@ -407,4 +407,23 @@
...
+
+  it "should close open producer sockets before they are GC'd" do
+    skip if defined?(JRUBY_VERSION)
+
+    @cleared = false
+    producer = rdkafka_config.producer
+    ObjectSpace.define_finalizer(producer, finalizer_proc)
+    producer = nil
+    # Request GC run.
+    RUBY_PLATFORM.casecmp?('java') ? java.lang.System.gc : GC.start
```
Comment (Adithya-copart) [inline spec/rdkafka/producer_spec.rb:425]: GC run is not guaranteed upon invocation as per the docs, so this test could be unreliable as-is.

The approach to validate the instance variable is not working in JRuby even though the finalizer is called.
https://github.com/jruby/jruby/issues/6353

**Comment by thijsc** [comment]
Let's merge the handle class, and rebase this then.

File: lib/rdkafka/error.rb [hunk 2]
```diff
@@ -57,4 +61,13 @@ def initialize(response, topic_partition_list, message_prefix=nil)
       @topic_partition_list = topic_partition_list
     end
   end
+
+  # Consumer closed error.
+  class ConsumerClosedError < BaseError; end
```
Comment (Adithya-copart) [inline lib/rdkafka/error.rb:66]: Relevant specs are needed.

**Comment by thijsc** [comment]
Could you rebase this one?

**Comment by thijsc** [comment]
Sorry for the delay. I've managed to switch CI providers and the build is now more stable. Could you rebase on `main`?

**Comment by thijsc** [comment]
Could you rebase this?

**Comment by thijsc** [comment]
@Adithya-copart Any chance you could rebase this one?

**Comment by thijsc** [comment]
Closing this one for the moment to get to a clean slate, please reopen when it can be merged :-)

**Pull request closed without merging**